egs/librispeech/ASR/transducer_stateless/model.py

```diff
@@ -102,24 +128,63 @@ def forward(
 
         decoder_out = self.decoder(sos_y_padded)
 
-        logits = self.joiner(encoder_out, decoder_out)
-
-        # rnnt_loss requires 0 padded targets
         # Note: y does not start with SOS
         y_padded = y.pad(mode="constant", padding_value=0)
 
-        assert hasattr(torchaudio.functional, "rnnt_loss"), (
-            f"Current torchaudio version: {torchaudio.__version__}\n"
-            "Please install a version >= 0.10.0"
+        boundary = torch.zeros(
+            (x.size(0), 4), dtype=torch.int64, device=x.device
+        )
+        boundary[:, 2] = y_lens
+        boundary[:, 3] = x_lens
+
+        # calculate prune ranges
+        simple_loss, (px_grad, py_grad) = k2.rnnt_loss_smoothed(
+            decoder_out,
+            encoder_out,
+            y_padded.to(torch.int64),
+            blank_id,
+            lm_only_scale=self.lm_scale,
+            am_only_scale=self.am_scale,
+            boundary=boundary,
+            return_grad=True,
+        )
+        ranges = k2.get_rnnt_prune_ranges(
+            px_grad, py_grad, boundary, self.prune_range
         )
 
-        loss = torchaudio.functional.rnnt_loss(
-            logits=logits,
-            targets=y_padded,
-            logit_lengths=x_lens,
-            target_lengths=y_lens,
-            blank=blank_id,
-            reduction="sum",
+        # forward loss
+        am_pruned, lm_pruned = k2.do_rnnt_pruning(
+            encoder_out, decoder_out, ranges
+        )
+        logits = self.joiner(am_pruned, lm_pruned)
+        pruned_loss = k2.rnnt_loss_pruned(
+            logits, y_padded.to(torch.int64), ranges, blank_id, boundary
         )
 
-        return loss
+        eos_y = add_eos(y, eos_id=blank_id)
+        eos_y_padded = eos_y.pad(mode="constant", padding_value=blank_id)
```

Comment:
... I'm not sure why we need this add_eos in RNN-T, especially since we're using blank and not real EOS.
In language modeling you need an extra EOS because it's predicted and it's part of the loss function, but
there isn't supposed to be any EOS in RNN-T's and I think the blank would just behave like extra unnecessary padding.
All this code is quite mysterious to me, IMO it needs documentation at each stage of what the matrices look like.